opentracing-spring-cloud/src/main/java/io/opentracing/contrib/spring/cloud/async/ExecutorBeanPostProcessor.java

```diff
@@ -0,0 +1,120 @@
+import org.springframework.util.ReflectionUtils;
+
+import io.opentracing.Tracer;
+import io.opentracing.contrib.spring.cloud.async.instrument.TracedExecutor;
+import io.opentracing.contrib.spring.cloud.async.instrument.TracedThreadPoolTaskExecutor;
+
+/**
+ * Bean post processor that wraps a call to an {@link Executor} either in a
+ * JDK or CGLIB proxy. Depending on whether the implementation has a final
+ * method or is final.
+ *
+ * @author Marcin Grzejszczak
+ * @since 1.1.4
+ */
+class ExecutorBeanPostProcessor implements BeanPostProcessor {
+
+    private final Tracer tracer;
+
+    ExecutorBeanPostProcessor(Tracer tracer) {
+        this.tracer = tracer;
+    }
+
+    @Override
+    public Object postProcessBeforeInitialization(Object bean, String beanName) throws BeansException {
+        return bean;
+    }
+
+    @Override
+    public Object postProcessAfterInitialization(Object bean, String beanName) throws BeansException {
+        if (bean instanceof Executor && !(bean instanceof ThreadPoolTaskExecutor)) {
+            Method execute = ReflectionUtils.findMethod(bean.getClass(), "execute", Runnable.class);
+            boolean methodFinal = Modifier.isFinal(execute.getModifiers());
+            boolean classFinal = Modifier.isFinal(bean.getClass().getModifiers());
+            boolean cglibProxy = !methodFinal && !classFinal;
+            Executor executor = (Executor) bean;
+            ProxyFactoryBean factory = new ProxyFactoryBean();
+            factory.setProxyTargetClass(cglibProxy);
+            factory.addAdvice(new ExecutorMethodInterceptor<>(executor, tracer));
+            factory.setTarget(bean);
+            return factory.getObject();
+        } else if (bean instanceof ThreadPoolTaskExecutor) {
+            boolean classFinal = Modifier.isFinal(bean.getClass().getModifiers());
+            boolean cglibProxy = !classFinal;
+            ThreadPoolTaskExecutor executor = (ThreadPoolTaskExecutor) bean;
+            ProxyFactoryBean factory = new ProxyFactoryBean();
+            factory.setProxyTargetClass(cglibProxy);
+            factory.addAdvice(new ExecutorMethodInterceptor<ThreadPoolTaskExecutor>(executor, tracer) {
+                @Override Executor tracedExecutor(Tracer tracer, ThreadPoolTaskExecutor executor) {
+                    return new TracedThreadPoolTaskExecutor(tracer, executor);
+                }
+            });
+            factory.setTarget(bean);
+            return factory.getObject();
+        }
+        return bean;
+    }
+}
+
+class ExecutorMethodInterceptor<T extends Executor> implements MethodInterceptor {
+
+    private final T delegate;
+    private final Tracer tracer;
+
+    ExecutorMethodInterceptor(T delegate, Tracer tracer) {
+        this.delegate = delegate;
+        this.tracer = tracer;
+    }
+
+    @Override
+    public Object invoke(MethodInvocation invocation) throws Throwable {
+        Executor tracedExecutor = tracedExecutor(tracer, delegate);
```

Comment:
Can the executor be obtained in the constructor, as the `tracedExecutor` is not dependent upon the `invocation`, but would result in a new `TracedExecutor` being created for each `invoke`.

Reply:
This is not being called frequently. It's bean post processing. Executors should be defined as singleton beans. This will create traced version per executor. (traced executor is a light-weight object)